### PR TITLE
Only aggregations require at least one shard request

### PR DESCRIPTION
**Pull request by piergm**
In the special case where we have no hits because the coordinator rewrote the query to match none we need to get at least one search response in order to produce a valid search result. This is true only if we have an aggregation in the query. 
Only for aggs we need to contact the shard because we need the mappings for the queried index in order to produce a valid aggregation response, and mappings are only materialized in memory in data nodes where a certain index is allocated. 

This pr updates this logic and sets `requireAtLeastOneMatch` only when searchRequest has aggregations.

The updated logic will set searchResponse.skippedShards == searchResponse.successfulShards == totalShards when we can skip all the shards because we have no aggs.

### CODE REVIEW

**Comment by piergm** [comment]
@elasticsearchmachine test this

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:1439]: I believe our code doesn't  _really_ ensure the that `searchRequest.source()` is always non-null. We check for null across the codebase. Shall we add a null check here before calling `aggregations()` ?

**Review by andreidan — APPROVED** [review]
Thanks for working on this Matteo. 

Really nice PR description and fix 🚀 

LGTM, assuming CI is happy

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java
Comment (piergm) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:1439]: Nice catch :D yes, I'll add that!
About CI I'll wait for multiple green runs before merging it :D

**Comment by piergm** [comment]
@elasticsearchmachine test this

**Comment by javanna** [comment]
Why is this a 9.0 only change? Should we backport it?

File: qa/multi-cluster-search/src/test/java/org/elasticsearch/search/CCSDuelIT.java
Comment (javanna) [inline qa/multi-cluster-search/src/test/java/org/elasticsearch/search/CCSDuelIT.java:584]: this is interesting, the previous suggested we were skipping can match, that would have been a better work-around, but if your code change was required, that means that we were not skipping can match? was it outdated then?

File: x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorCanMatchIT.java
Comment (javanna) [inline x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorCanMatchIT.java:210]: Interesting, why was it null before? We were unskipping one shard, yet getting null total hits?

**Review by javanna — COMMENTED** [review]
I left a couple of questions, this is a good cleanup!

**Comment by piergm** [comment]
> Why is this a 9.0 only change? Should we backport it?

Thanks for the review @javanna, my reasoning here is that this is not a bug therefore I wasn't sure if we need to backport it. Perhaps just to 8.x?

File: qa/multi-cluster-search/src/test/java/org/elasticsearch/search/CCSDuelIT.java
Comment (piergm) [reply qa/multi-cluster-search/src/test/java/org/elasticsearch/search/CCSDuelIT.java:584]: The previous comment was outdated, there was a refactor of this method a while back and the `setPreFilterShardSize` the comments referees to was removed

File: x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorCanMatchIT.java
Comment (piergm) [reply x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorCanMatchIT.java:210]: It's because of SearchPhaseController#merge when we skip all the shards we have an empty result and therefore we `return SearchResponseSections.EMPTY_WITH_TOTAL_HITS;`. I wonder if this should instead be `EMPTY_WITHOUT_TOTAL_HITS`.

**Comment by piergm** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/115794)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by javanna** [comment]
Yes, we tend to backport changes to 8.x for now (next 8.x minor release), bugfixes only to upcoming patch releases.

**Comment by cbuescher** [comment]
I wanted to backport this change but apparently the original PR that create the test issue never got merged to 8x (contrary to what the labeling says). https://github.com/elastic/elasticsearch/pull/115794
I'll refrain from backporting for now because of this.

**Comment by javanna** [comment]
Found the same inconsistency around backporting. I think that this change is mislabelled, I will relabel 8.18.0.

**Pull request merged**
